oak_functions/wasm/src/lib.rs

```diff
@@ -36,388 +36,470 @@ use oak_functions_abi::{
+
+        let invoke_extension = wasmi::Func::wrap(
+            &mut store,
+            // The types in the signatures correspond to the parameters from
+            // oak_functions_abi/src/lib.rs.
+            |mut caller: wasmi::Caller<'_, UserState<L>>,
+             handle: AbiExtensionHandle,
+             request_ptr: AbiPointer,
+             request_len: AbiPointerOffset,
+             response_ptr_ptr: AbiPointer,
+             response_len_ptr: AbiPointer| {
+                let result = invoke_extension(
+                    &mut caller,
+                    handle,
+                    request_ptr,
+                    request_len,
+                    response_ptr_ptr,
+                    response_len_ptr,
                 );
-                OakStatus::ErrInvalidArgs
-            })?;
-        self.response_bytes = response;
-        Ok(())
-    }
 
-    pub fn invoke_extension(
-        &mut self,
-        handle: AbiExtensionHandle,
-        request_ptr: AbiPointer,
-        request_len: AbiPointerOffset,
-        response_ptr_ptr: AbiPointer,
-        response_len_ptr: AbiPointer,
-    ) -> Result<(), OakStatus> {
-        let handle: ExtensionHandle = ExtensionHandle::from_i32(handle).ok_or_else(|| {
-            self.log_error(&format!("couldn't convert handle {:?} from i32", handle));
-            OakStatus::ErrInvalidHandle
-        })?;
+                from_oak_status_result(result)
+            },
+        );
 
-        let request = self
-            .read_buffer_from_wasm_memory(request_ptr, request_len)
+        linker
+            .define(HOST, "invoke", invoke_extension)
+            .expect("failed to define invoke in linker");
+
+        // Use linker and store to get instance of module.
+        let instance = linker
+            .instantiate(&mut store, &module)
+            .map_err(|err| anyhow::anyhow!("failed to instantiate Wasm module: {:?}", err))?
+            // Use `main` as entry point.
+            .ensure_no_start(&mut store)
             .map_err(|err| {
-                self.log_error(&format!(
-                    "handle {:?}: unable to read input from guest memory: {:?}",
-                    handle, err
-                ));
-                OakStatus::ErrInvalidArgs
+                anyhow::anyhow!("failed to ensure no start in Wasm module: {:?}", err)
             })?;
 
-        let extension = match self.extensions.get_mut(&handle) {
-            // Can't convince the borrow checker to use `ok_or_else` to `self.log_error`.
-            Some(extension) => Ok(extension),
-            None => {
-                self.log_error(&format!("cannot find extension with handle {:?}", handle));
-                Err(OakStatus::ErrInvalidHandle)
-            }
-        }?;
-
-        let response = extension.invoke(request)?;
-        self.alloc_and_write_buffer_to_wasm_memory(response, response_ptr_ptr, response_len_ptr)
+        // Check that the instance exports "alloc".
```

Comment:
As a future todo, it may be nice to list the [exports](https://docs.rs/wasmi/0.28.0/wasmi/struct.Instance.html#method.exports) from the instance, to help debugging.

Reply:
Added TODO(#3785)